canary/S3ObjectTransport.cpp

```diff
@@ -157,40 +336,434 @@ void S3ObjectTransport::GetObject(
+                requestOptions.onStreamComplete =
+                    [uploadId, keyPath, signedRequest, conn, completedCallback](Http::HttpStream &stream, int error) {
+                        int errorCode = error;
+
+                        if (!errorCode)
+                        {
+                            errorCode = (stream.GetResponseStatusCode() == 200) ? AWS_OP_SUCCESS : AWS_OP_ERR;
+
+                            aws_log_level logLevel = (errorCode == AWS_OP_ERR) ? AWS_LL_ERROR : AWS_LL_INFO;
+
+                            AWS_LOGF(
+                                logLevel,
+                                AWS_LS_CRT_CPP_CANARY,
+                                "Created multipart upload for path %s with response status %d.",
+                                keyPath.c_str(),
+                                stream.GetResponseStatusCode());
+                        }
+                        else
+                        {
+                            AWS_LOGF_ERROR(
+                                AWS_LS_CRT_CPP_CANARY,
+                                "Completed multipart upload for path %s with error '%s'",
+                                keyPath.c_str(),
+                                aws_error_debug_str(errorCode));
+                        }
+
+                        if (uploadId->empty())
+                        {
+                            errorCode = AWS_OP_ERR;
+                        }
+
+                        completedCallback(errorCode, uploadId);
+                    };
+
+                conn->NewClientStream(requestOptions);
+            });
+        });
+}
+
+void S3ObjectTransport::CompleteMultipartUpload(
+    const Aws::Crt::String &key,
+    const Aws::Crt::String &uploadId,
+    const Aws::Crt::Vector<Aws::Crt::String> &etags,
+    CompleteMultipartUploadCompleted completedCallback)
+{
+    AWS_LOGF_INFO(AWS_LS_CRT_CPP_CANARY, "Completing multipart upload for %s...", key.c_str());
+
+    auto completeMultipartUploadRequest = MakeShared<Http::HttpRequest>(g_allocator, g_allocator);
+    completeMultipartUploadRequest->AddHeader(m_hostHeader);
+    completeMultipartUploadRequest->SetMethod(aws_http_method_post);
+
+    std::shared_ptr<StringStream> xmlContents = std::make_shared<StringStream>();
+    *xmlContents << "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n";
+    *xmlContents << "<CompleteMultipartUpload xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">\n";
+
+    for (size_t i = 0; i < etags.size(); ++i)
+    {
+        const Aws::Crt::String &etag = etags[i];
+        size_t partNumber = i + 1;
+
+        *xmlContents << "   <Part>\n";
+        *xmlContents << "       <ETag>" << etag << "</ETag>\n";
+        *xmlContents << "       <PartNumber>" << partNumber << "</PartNumber>\n";
+        *xmlContents << "   </Part>\n";
+    }
+
+    *xmlContents << "</CompleteMultipartUpload>";
+
+    aws_input_stream *inputStream = Aws::Crt::Io::AwsInputStreamNewCpp(xmlContents, aws_default_allocator());
```

Comment:
My understanding of the aws_http_message_set_body_stream API was that it took ownership of the memory and handled cleaning it up.  I believe, although I don't have it on hand, I did find the place where it de-allocates it.  I'll double check my recall on this one and respond once I find that place in code. :)

Reply:
Hey, it looks like I'm wrong--I'll fix that and the other occurrences of it.  Apologies for that!

Reply:
So the ownership of the stream gets passed to the request during the aws_http_message_set_body_stream.  The request is of type HttpResponse which inherits from HttpMessage.  The destructor of HttpMessage deletes the stream.  Short term, I'll add a comment to these places so that this obvious.  Maybe long term we can make an API change?